### ISSUE TITLE
Add tagging separate from string descriptions

### ISSUE THREAD

**Issue by thebigmunch** [darrenburns/ward#107]
There is a clear distinction between describing a test and classifying it. They are semantically different things. The current method of 'tagging' or 'marking' tests is by patterns in the string descriptions. This not only mixes two things that probably shouldn't be but also causes some usability issues.

Even just a couple 'tags' can drown out the test descriptions, especially since they have to be formatted for reliable parsing for text search. They don't end up looking great in the output as the tags don't line up for easy human parsing. As alluded to before, text searching is clunky at best or unreliable.

So, I suggest adding a ``tags`` or similar attribute to tests and parameter to the ``test`` decorator to supplant the idea of tagging the descriptions. The ``--search`` CLI option could remain for text searching the descriptions, but a new ``--tags`` (or whatever name is chosen) option could be added. An implementation detail that would need to be ironed out is whether to support both 'or' and 'and' cases as well as how for multiple tags.

**Comment by darrenburns** [comment]
Thanks for the suggestion. I totally agree with it, and it's something I've had on my mind and discussed with people in the past. Suggested names for the concept included `groups`, `tags`, and `labels`. Personally I think `tags` is my favourite.

```python
@test("1 + 2 == 3", tags=["apples", "bananas"])
def _():
	assert 1 + 2 == 3
```

From the command line POV, *maybe* an expression like those below would work, but perhaps it's not extensible enough? It does leave open some flexibility around the `and` vs `or` situation, and how they can be queried in general. It's hard to know without having very concrete use-cases in mind in advance.

```
ward --tags "apples and bananas"
```

```
ward --tags "apples or bananas"
```

**Comment by thebigmunch** [comment]
> It's hard to know without having very concrete use-cases in mind in advance.

So, for an example of an ``and`` use-case, tag unit and integration tests as such then using any other tag. I have an audio metadata library where I often want to do something like: "run unit tests for FLAC" ("unit and FLAC"). An example using this library for the ``or`` use-case might be when I want to run tests just for certain formats, e.g. "run tests for MP3 and ID3" ("MP3 or ID3"). So there is a starting point as far as real-world examples.

> Regarding the output, I don't think tags would need to be included in the output by default and, as you said, that'd leave more room for the test description.

Agreed.

**Comment by thebigmunch** [comment]
So, as for a CLI for this, one way might be this:

```
ward --tags unit,FLAC --tags MP3 --tags ID3
```

Matching my examples above, this would mean run tests tagged as ``unit and FLAC`` or tagged as ``MP3`` or tagged as ``ID3``.

Splitting on commas, this would result in the tags looking like:

``` python
[
	['unit', 'FLAC'],
	['MP3'],
	['ID3'],
]
```

The code to check if a test matches the supplied tags would be something like:

``` python
any(
	all(
		tag in test.tags
		for tag in tag_set
	)
	for tag_set in tags
)
```

**Comment by darrenburns** [comment]
I done some further research into this, and there's a few test frameworks that already support the concept of tags. They tend to have an API like:

`ward --tag unit --tag FLAC --exclude-tag integration`

i.e. query all tests that are tagged with `unit`, **or** tagged with `FLAC` (or both), and are **not** tagged with `integration`. Quite similar to your suggestion @thebigmunch, except it doesn't have any way of representing an "and". I'm not sure about the comma separated tags via the CLI to represent the conjunction. Something I'll sleep on :)

Some frameworks I looked at:
* **Django**: https://docs.djangoproject.com/en/3.0/topics/testing/tools/#tagging-tests
* **ScalaTest**: http://www.scalatest.org/user_guide/using_the_runner#specifyingTagsToIncludeAndExclude

**Comment by thebigmunch** [comment]
> I'm not sure about the comma separated tags via the CLI to represent the conjunction

* I can picture ``+`` being used in tags common enough to want to spare it. And I don't think it looks as good.
* ``/`` is even more likely to be used than ``+``.
* ``&`` would require shell escaping, which isn't great. That's one of the reasons I dislike your initial syntax suggestion.
* This is actually close to the use of the comma in language, so I don't think it will be strange for people. Most will understand it quickly from just a bit of help text.

CLI syntax is too simple for an absolutely perfect analogue for this complexity, which isn't even *that* complex. So, a bit of stretching is going to be necessary. I'd love to hear others' thoughts, opinions, or suggestions.

I haven't really had much need for an exclude, but I can definitely see that being desired. That could easily share the same design as the ``--tags`` option with an added condition like:

``` python
and not any(
	all(
		tag in test.tags
		for tag in tag_set
	)
	for tag_set in exclude_tags
)
```

**Comment by onlyanegg** [comment]
I like this, and was thinking about it earlier. Another use case could be tagging long running tests in order to skip them. It might also be nice to add an interface to tag an entire module with a set of tags, for example, if all tests in the module are unit tests.

**Comment by thebigmunch** [comment]
So, I was playing around on lunch break and did this: https://github.com/thebigmunch/ward/tree/tags. Mainly just so I could play around with and get a feel for the idea. Pushed it so others could as well. It uses the syntax I suggested, but that could be changed pretty easily if anyone wants to get a feel for different syntax suggestions.

PS I definitely could've missed something somewhere; I haven't had a chance to get particularly familiar with the code yet. And this was all done in a short lunch break, so I can't promise I didn't make a mistake somewhere. But it seems to work : P

**Comment by darrenburns** [comment]
I'll take a deeper look, but I had a quick skim. Only comment I have for now is that when I mentioned `exclude_tags`, it wasn't an extra argument for the `@test` decorator, it was a way of specifying which tags should be excluded from the run :)

**Comment by thebigmunch** [comment]
> Only comment I have for now is that when I mentioned `exclude_tags`, it wasn't an extra argument for the `@test` decorator, it was a way of specifying which tags should be excluded from the run :)

Yeah, that's definitely just a case of me being in a hurry to get it done in the time I had. That definitely shouldn't have been there; it makes no sense and isn't actually used for anything. I've force-pushed the removal and separated the exclude_tags check into its own variable so it's less of a wall.

**Comment by thebigmunch** [comment]
So, I finally came back to this and created a branch implementing cucumber tag expressions in [this branch](https://github.com/thebigmunch/ward/tree/tags-cucumber).

I actually thought of a syntax I use somewhere else to use the style in my tags branch without the (exclude-)tags options being multiple. But, I think the advantages of the cucumber tag expressions outweigh having to quote them:

* They are maintained/documented by someone else (super plus).
  Caveat: it's documented as using ``@`` to denote the tags, but that isn't actually supported. So, their documentation is wrong?
* They support far more complex scenarios for those edge case users.
* They don't require multiple-style options.
* Did I mention the parser doesn't need to be maintained/documented by any of us? ^_^

**Comment by darrenburns** [comment]
Ooooh, very nice @thebigmunch! Would be brilliant to get this feature into Ward. Would you like to open a draft PR and we can discuss moving this forward in there?

**Comment by darrenburns** [comment]
Tagging tests and querying those tags using tag expressions has been released as part of Ward 0.42.0b0, which now on PyPI.
